README.md
# Future Local Storage

An init-once-per-future cell for thread-local values.

<!-- ANCHOR: description -->

This crate provides an [`FutureOnceCell`] cell-like type, which provides the
similar API as the [`tokio::task_local`] but without using any macros.

Future local storage associates a value to the context of a given future. After
the future finished it returns this value back to the caller. That meaning that
the values is passed through the context of the executed future. This
functionality can be useful for tracing async code or adding metrics to it.

## Usage

```rust
use std::cell::Cell;

use future_local_storage::FutureOnceCell;

static VALUE: FutureOnceCell<Cell<u64>> = FutureOnceCell::new();

#[tokio::main]
async fn main() {
    let (output, answer) = VALUE.scope(Cell::from(0), async {
        VALUE.with(|x| {
            let value = x.get();
            x.set(value + 1);
        });

        "42".to_owned()
    }).await;

    assert_eq!(output.into_inner(), 1);
    assert_eq!(answer, "42");
}
```

[`tokio::task_local`]: https://docs.rs/tokio/latest/tokio/macro.task_local.html

<!-- ANCHOR_END: description -->

[`FutureOnceCell`]: #Usage


CHANGELOG
# Changelog

## Unreleased

-

## 0.1.2

- Use `include_md` macro to make documentation more readable.

## 0.1.1

- Fix licence in crate and CI rules.

## 0.1.0

First public release


src/imp.rs
//! Low-level implementation details.

use std::{cell::RefCell, fmt::Debug};

use state::LocalInitCell;

/// A type wrapper that provides interior mutability and allows for safe and efficient access to an
/// optional value stored in a cell.
pub type LocalKey<T> = RefCell<Option<T>>;

/// A future local storage key which owns its content.
///
/// It uses thread local storage to ensure that the each polled future has its own local storage key.
pub struct FutureLocalKey<T>(LocalInitCell<LocalKey<T>>);

impl<T> FutureLocalKey<T> {
    /// Creates an empty future local key.
    #[inline]
    pub const fn new() -> Self {
        Self(LocalInitCell::new())
    }
}

impl<T: Send + 'static> FutureLocalKey<T> {
    /// Returns a reference to the underlying thread local storage key, and if it has not been initalized,
    /// initializes it with the `None` value.
    ///
    /// # Important
    ///
    /// Using this method ensures that the local key is initialized, use only it ot access the underlying
    /// thread local key.
    #[inline]
    pub fn local_key(&'static self) -> &'static LocalKey<T> {
        self.0.set(|| RefCell::new(None));
        self.0.get()
    }

    /// Swaps the underlying value and the given one, without deinitializing either one.
    #[inline]
    pub fn swap(this: &'static Self, other: &mut Option<T>) {
        std::mem::swap(other, &mut *this.local_key().borrow_mut());
    }
}

impl<T: Send + 'static + Debug> Debug for FutureLocalKey<T> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_tuple("FutureLocalKey").field(&self.0).finish()
    }
}

#[cfg(test)]
mod tests {
    use std::{cell::Cell, thread::JoinHandle};

    use pretty_assertions::assert_eq;

    use super::*;

    #[test]
    fn test_future_local_key_init() {
        static KEY: FutureLocalKey<i32> = FutureLocalKey::new();

        assert_eq!(*KEY.local_key().borrow(), None);
    }

    #[test]
    fn test_future_local_key_swap() {
        static KEY: FutureLocalKey<String> = FutureLocalKey::new();

        let threads = (0..42).map(|i| {
            std::thread::spawn(move || {
                let mut slot = Some(i.to_string());
                // Swap keys and make sure that the slot and key content actually has been swapped.
                FutureLocalKey::swap(&KEY, &mut slot);
                assert_eq!(slot, None);
                assert_eq!(*KEY.local_key().borrow(), Some(i.to_string()));

                // Swap keys again.
                FutureLocalKey::swap(&KEY, &mut slot);
                assert_eq!(slot, Some(i.to_string()));
                assert_eq!(*KEY.local_key().borrow(), None);
            })
        });

        threads.into_iter().try_for_each(JoinHandle::join).unwrap();
    }

    // Test [`state::LocalInitCell`] itself.
    #[test]
    fn test_local_init_cell_multiple_threads() {
        static VALUE: LocalInitCell<Cell<usize>> = LocalInitCell::new();
        VALUE.set(|| Cell::new(0));

        let handle = std::thread::spawn(|| {
            assert_eq!(VALUE.get().get(), 0);
            VALUE.get().set(2);
            assert_eq!(VALUE.get().get(), 2);
        });

        assert_eq!(VALUE.get().get(), 0);
        handle.join().unwrap();
        // Make sure that after the thread will be finished, the value will not change.
        assert_eq!(VALUE.get().get(), 0);
    }
}


src/old.rs
use std::{
    cell::{Ref, RefCell},
    future::Future,
    pin::Pin,
    task::{Context, Poll},
};

use pin_project_lite::pin_project;
use state::LocalInitCell;

type LocalKey<T> = RefCell<Option<T>>;

pub struct FutureLock<T>(LocalInitCell<LocalKey<T>>);

impl<T: Send> FutureLock<T> {
    pub const fn new() -> Self {
        Self(LocalInitCell::new())
    }

    fn local_key(&'static self) -> &LocalKey<T> {
        if self.0.try_get().is_none() {
            self.0.set(|| RefCell::new(None));
        }

        self.0.get()
    }

    fn get(&'static self) -> Ref<Option<T>> {
        let local_key = self.local_key();
        let value = local_key.borrow();
        value
    }

    fn attach<F: Future>(&'static self, fut: F) -> InstrumentedFuture<T, F> {
        InstrumentedFuture {
            inner: fut,
            future_lock: self,
            stored_value: None,
        }
    }

    pub fn set(&'static self, value: T) {
        self.local_key().borrow_mut().replace(value);
    }

    pub fn with_or_init<F, R, I>(&'static self, mut with: F, init: I) -> R
    where
        F: FnMut(&T) -> R,
        I: FnOnce() -> T,
    {
        if self.get().is_none() {
            self.set(init());
        }

        with(self.local_key().borrow().as_ref().unwrap())
    }
}

impl<T> FutureLock<T>
where
    T: Copy + Send,
{
    #[must_use]
    pub fn get_or_init<I>(&'static self, init: I) -> T
    where
        I: FnOnce() -> T,
    {
        if self.get().is_none() {
            let value = init();
            self.set(value);
            return value;
        }

        self.local_key().borrow().unwrap()
    }
}

pin_project! {
    pub struct InstrumentedFuture<T: 'static, F> {
        // TODO add support to instrument Drop.
        #[pin]
        inner: F,
        future_lock: & 'static FutureLock<T>,
        stored_value: Option<T>,
    }
}

impl<T, F> Future for InstrumentedFuture<T, F>
where
    T: Send,
    F: Future,
{
    type Output = F::Output;

    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
        let this = self.project();
        // Swap in future local key.
        std::mem::swap(
            this.stored_value,
            &mut *this.future_lock.local_key().borrow_mut(),
        );
        // Poll the underlying future.
        let result = this.inner.poll(cx);
        // Swap future local key back.
        std::mem::swap(
            this.stored_value,
            &mut *this.future_lock.local_key().borrow_mut(),
        );

        result
    }
}

pub struct FutureLazy<T> {
    lock: FutureLock<T>,
    init: fn() -> T,
}

impl<T: Send> FutureLazy<T> {
    pub const fn new(init: fn() -> T) -> Self {
        Self {
            lock: FutureLock::new(),
            init,
        }
    }

    pub fn with<F, R>(&'static self, with: F) -> R
    where
        F: FnMut(&T) -> R,
    {
        self.lock.with_or_init(with, self.init)
    }

    pub fn set(&'static self, value: T) {
        self.lock.set(value)
    }
}

impl<T> FutureLazy<T>
where
    T: Copy + Send,
{
    pub fn get(&'static self) -> T {
        self.with(|x| *x)
    }
}

pub trait Storage<T>: private::Sealed {
    fn attach<F: Future>(&'static self, fut: F) -> InstrumentedFuture<T, F>;
}

impl<T: Send> Storage<T> for FutureLock<T> {
    fn attach<F: Future>(&'static self, fut: F) -> InstrumentedFuture<T, F> {
        InstrumentedFuture {
            inner: fut,
            future_lock: self,
            stored_value: None,
        }
    }
}

impl<T: Send> Storage<T> for FutureLazy<T> {
    fn attach<F: Future>(&'static self, fut: F) -> InstrumentedFuture<T, F> {
        self.lock.attach(fut)
    }
}

pub trait FutureLocalStorage: Sized + private::Sealed {
    fn attach<T: Send, S: Storage<T>>(self, lock: &'static S) -> InstrumentedFuture<T, Self>;
}

impl<F: Future> FutureLocalStorage for F {
    fn attach<T: Send, S: Storage<T>>(self, lock: &'static S) -> InstrumentedFuture<T, Self> {
        lock.attach(self)
    }
}

mod private {
    use std::future::Future;

    use super::{FutureLazy, FutureLock};

    pub trait Sealed {}

    impl<F: Future> Sealed for F {}
    impl<T> Sealed for FutureLock<T> {}
    impl<T> Sealed for FutureLazy<T> {}
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_get_set_trivial() {
        static VALUE: FutureLock<u64> = FutureLock::new();

        assert_eq!(*VALUE.get(), None);
        VALUE.set(42);

        assert_eq!(*VALUE.get(), Some(42));
    }

    #[test]
    fn test_with_init() {
        static VALUE: FutureLock<String> = FutureLock::new();

        let s = VALUE.with_or_init(|s| s.clone(), || "Back to the future".to_owned());
        assert_eq!(s, "Back to the future");
    }

    #[test]
    fn test_set_multiple_threads() {
        static VALUE: FutureLock<u64> = FutureLock::new();
        VALUE.set(1);

        let handle = std::thread::spawn(|| {
            assert_eq!(*VALUE.get(), None);
            VALUE.set(2);
            assert_eq!(*VALUE.get(), Some(2));
        });

        assert_eq!(*VALUE.get(), Some(1));
        handle.join().unwrap();

        assert_eq!(*VALUE.get(), Some(1));
    }

    #[tokio::test]
    async fn test_future_lock_attach() {
        static VALUE: FutureLock<u64> = FutureLock::new();

        let fut_1 = async {
            for _ in 0..42 {
                let j = VALUE.get_or_init(|| 0);
                VALUE.set(j + 1);
                tokio::task::yield_now().await;
            }

            VALUE.get().unwrap()
        }
        .attach(&VALUE);

        let fut_2 = async { VALUE.with_or_init(|x| *x, || 15) }.attach(&VALUE);

        assert_eq!(fut_1.await, 42);
        assert_eq!(fut_2.await, 15);
        assert_eq!(
            tokio::spawn(async { VALUE.get_or_init(|| 115) }.attach(&VALUE))
                .await
                .unwrap(),
            115
        );
    }

    #[tokio::test]
    async fn test_future_lazy_attach() {
        static VALUE: FutureLazy<i32> = FutureLazy::new(|| -1);

        let fut_1 = async {
            for _ in 0..42 {
                let j = VALUE.with(|x| *x);
                VALUE.set(j + 1);
                 tokio::task::yield_now().await;
            }

            VALUE.get()
        }
        .attach(&VALUE);

        let fut_2 = async {
            VALUE.set(15);
             tokio::task::yield_now().await;
            VALUE.get()
        }
        .attach(&VALUE);

        assert_eq!(fut_1.await, 41);
        assert_eq!(fut_2.await, 15);
        assert_eq!(
            tokio::spawn(
                async {
                    VALUE.set(115);
                     tokio::task::yield_now().await;
                    VALUE.get()
                }
                .attach(&VALUE)
            )
            .await
            .unwrap(),
            115
        );
    }
}


src/future.rs
//! Future types.

use std::{
    future::Future,
    pin::Pin,
    task::{Context, Poll},
};

use pin_project::{pin_project, pinned_drop};

use crate::{imp::FutureLocalKey, FutureLocalStorage};

impl<F: Future> FutureLocalStorage for F {
    fn with_scope<T, S>(self, scope: &'static S, value: T) -> ScopedFutureWithValue<T, Self>
    where
        T: Send,
        S: AsRef<FutureLocalKey<T>>,
    {
        let scope = scope.as_ref();
        ScopedFutureWithValue {
            inner: self,
            scope,
            value: Some(value),
        }
    }
}

/// A [`Future`] that sets a value `T` of a future local for the future `F` during its execution.
/// Unlike the [`ScopedFutureWithValue`] this future discards the future local value.
#[pin_project]
#[derive(Debug)]
pub struct ScopedFuture<T, F>(#[pin] ScopedFutureWithValue<T, F>)
where
    T: Send + 'static,
    F: Future;

impl<T, F> Future for ScopedFuture<T, F>
where
    T: Send,
    F: Future,
{
    type Output = F::Output;

    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
        self.project().0.poll(cx).map(|(_value, result)| result)
    }
}

impl<T, F> ScopedFutureWithValue<T, F>
where
    T: Send,
    F: Future,
{
    /// Discards the future local value from the future output.
    pub fn discard_value(self) -> ScopedFuture<T, F> {
        ScopedFuture(self)
    }
}

/// A [`Future`] that sets a value `T` of a future local for the future `F` during its execution.
///
/// This future also returns a future local value after execution.
#[pin_project(PinnedDrop)]
#[derive(Debug)]
pub struct ScopedFutureWithValue<T, F>
where
    T: Send + 'static,
    F: Future,
{
    // TODO Implement manually drop to provide scope access to the future Drop.
    #[pin]
    inner: F,
    scope: &'static FutureLocalKey<T>,
    value: Option<T>,
}

#[pinned_drop]
impl<T, F> PinnedDrop for ScopedFutureWithValue<T, F>
where
    F: Future,
    T: Send + 'static,
{
    fn drop(self: Pin<&mut Self>) {}
}

impl<T, F> Future for ScopedFutureWithValue<T, F>
where
    T: Send,
    F: Future,
{
    type Output = (T, F::Output);

    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
        let this = self.project();
        // Swap in future local key.
        FutureLocalKey::swap(this.scope, this.value);
        // Poll the underlying future.
        let result = this.inner.poll(cx);
        // Swap future local key back.
        FutureLocalKey::swap(this.scope, this.value);

        let result = std::task::ready!(result);
        // Take the scoped value to return it back to the future caller.
        let value = this.value.take().unwrap();
        Poll::Ready((value, result))
    }
}

impl<T, F> From<ScopedFutureWithValue<T, F>> for ScopedFuture<T, F>
where
    T: Send,
    F: Future,
{
    fn from(value: ScopedFutureWithValue<T, F>) -> Self {
        Self(value)
    }
}


src/lazy_lock.rs
use std::fmt::Debug;

use crate::imp::{self, FutureLocalKey};

/// A value which is initialized on the first access.
///
/// Similar to [`std::cell::LazyCell`], but local to a certain future.
pub struct FutureLazyLock<T> {
    inner: imp::FutureLocalKey<T>,
    // TODO Rewrite on top of unsafe cell.
    init: fn() -> T,
}

impl<T> FutureLazyLock<T> {
    /// Creates an empty future lazy lock.
    #[inline]
    pub const fn new(init: fn() -> T) -> Self {
        Self {
            inner: imp::FutureLocalKey::new(),
            init,
        }
    }
}

impl<T: Send + 'static> FutureLazyLock<T> {
    /// Returns a reference to a local key, initalizing it with the `init` if it has not been
    /// previously initialized.
    #[inline]
    fn inited_local_key(&'static self) -> &'static imp::LocalKey<T> {
        // Local key is empty only before init, so this branch runs only once.
        if !self.inner.local_key().borrow().is_some() {
            let mut value = Some((self.init)());
            imp::FutureLocalKey::swap(&self.inner, &mut value);
        }
        self.inner.local_key()
    }

    /// Acquires a reference to the value stored in this future local storage.
    ///
    /// This will lazy initialize value if the future has not referenced this key yet.
    #[inline]
    pub fn with<F, R>(&'static self, f: F) -> R
    where
        F: FnOnce(&T) -> R,
    {
        let value = self.inited_local_key().borrow();
        f(value.as_ref().unwrap())
    }

    /// Replaces a value stored in this future local storage by the given one and returns the
    /// previously stored value.
    ///
    /// This will lazy initialize value if the future has not referenced this key yet.
    #[inline]
    pub fn replace(&'static self, value: T) -> T {
        self.inited_local_key().borrow_mut().replace(value).unwrap()
    }

    /// Sets or initializes the contained value.
    ///
    /// Unlike the other methods, this will not run the lazy initializer of this storage.
    #[inline]
    pub fn set(&'static self, value: T) {
        self.inited_local_key().borrow_mut().replace(value);
    }

    /// Returns a copy of the contained value.
    ///
    /// This will lazy initialize value if the future has not referenced this key yet.
    #[inline]
    pub fn get(&'static self) -> T
    where
        T: Copy,
    {
        self.with(|x| *x)
    }
}

impl<T: Debug + Send + 'static> Debug for FutureLazyLock<T> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_tuple("FutureLazyLock").field(&self.inner).finish()
    }
}

impl<T> AsRef<FutureLocalKey<T>> for FutureLazyLock<T> {
    fn as_ref(&self) -> &FutureLocalKey<T> {
        &self.inner
    }
}

#[cfg(test)]
mod tests {
    use crate::FutureLocalStorage;

    use super::*;

    use pretty_assertions::assert_eq;

    #[test]
    fn test_lazy_lock_trivial() {
        static LOCK: FutureLazyLock<&str> = FutureLazyLock::new(|| "42");

        assert_eq!(LOCK.with(|x| *x), "42");
        LOCK.replace("abacaba");
        assert_eq!(LOCK.get(), "abacaba");
    }

    #[test]
    fn test_lazy_lock_multiple_threads() {
        static VALUE: FutureLazyLock<u64> = FutureLazyLock::new(|| 1);

        let handle = std::thread::spawn(|| {
            assert_eq!(VALUE.get(), 1);
            VALUE.set(2);
            assert_eq!(VALUE.get(), 2);
        });

        assert_eq!(VALUE.get(), 1);
        handle.join().unwrap();
        // Make sure that after the thread will be finished, the value will not change.
        assert_eq!(VALUE.get(), 1);
    }

    #[tokio::test]
    async fn test_future_lazy() {
        static VALUE: FutureLazyLock<i32> = FutureLazyLock::new(|| -1);

        let fut_1 = async {
            for _ in 0..42 {
                let j = VALUE.with(|x| *x);
                VALUE.replace(j + 1);
                tokio::task::yield_now().await;
            }

            VALUE.get()
        }
        .with_scope(&VALUE);

        let fut_2 = async {
            VALUE.replace(15);
            tokio::task::yield_now().await;
            VALUE.get()
        }
        .with_scope(&VALUE);

        assert_eq!(fut_1.await, 41);
        assert_eq!(fut_2.await, 15);
        assert_eq!(
            tokio::spawn(
                async {
                    VALUE.replace(115);
                    tokio::task::yield_now().await;
                    VALUE.get()
                }
                .with_scope(&VALUE)
            )
            .await
            .unwrap(),
            115
        );
    }
}


src/once_lock.rs
use std::{fmt::Debug, future::Future};

use crate::{
    imp::{self, FutureLocalKey},
    FutureLocalStorage, InstrumentedFuture,
};

pub struct FutureOnceLock<T>(imp::FutureLocalKey<T>);

impl<T> FutureOnceLock<T> {
    /// Creates an empty future once lock.
    #[must_use]
    pub const fn new() -> Self {
        Self(imp::FutureLocalKey::new())
    }
}

impl<T: Send + 'static> FutureOnceLock<T> {
    /// Acquires a reference to the value in this future local storage.
    ///
    /// Unlike the [`std::thread::LocalKey::with`] this method does not initialize the value
    /// when called.
    ///
    /// # Panics
    ///
    /// This method will panic if the future local doesn't have a value set.
    #[inline]
    pub fn with<F, R>(&'static self, f: F) -> R
    where
        F: FnOnce(&T) -> R,
    {
        let value = self.0.local_key().borrow();
        f(value
            .as_ref()
            .expect("cannot access a future local value without setting it first"))
    }

    #[inline]
    pub fn take(&'static self) -> Option<T> {
        self.0.local_key().borrow_mut().take()
    }

    #[inline]
    pub fn get(&'static self) -> Option<T>
    where
        T: Copy,
    {
        *self.0.local_key().borrow()
    }

    /// Sets a value `T` as the future-local value for the future `F`.
    ///
    /// On completion of `scope`, the future-local value will be dropped.
    #[inline]
    pub fn scope<F>(&'static self, value: T, future: F) -> InstrumentedFuture<T, F>
    where
        F: Future,
    {
        future.with_scope(self, value)
    }
}

impl<T: Debug + Send + 'static> Debug for FutureOnceLock<T> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_tuple("FutureOnceLock").field(&self.0).finish()
    }
}

impl<T> AsRef<FutureLocalKey<T>> for FutureOnceLock<T> {
    fn as_ref(&self) -> &FutureLocalKey<T> {
        &self.0
    }
}

#[cfg(test)]
mod tests {
    use pretty_assertions::assert_eq;

    use crate::FutureLocalStorage;

    use super::*;

    impl<T: Send + 'static> FutureOnceLock<T> {
        fn replace(&'static self, value: T) -> Option<T> {
            self.0.local_key().borrow_mut().replace(value)
        }

        fn set(&'static self, value: T) {
            self.replace(value);
        }
    }

    #[test]
    fn test_once_lock_trivial() {
        static LOCK: FutureOnceLock<String> = FutureOnceLock::new();
        LOCK.set("0".to_owned());

        assert_eq!(LOCK.with(Clone::clone), "0".to_owned());
        LOCK.set("42".to_owned());
        assert_eq!(LOCK.with(Clone::clone), "42".to_owned());
    }

    #[test]
    fn test_once_lock_multiple_threads() {
        static VALUE: FutureOnceLock<u64> = FutureOnceLock::new();
        VALUE.replace(1);

        let handle = std::thread::spawn(|| {
            assert_eq!(VALUE.get(), None);
            VALUE.replace(2);
            assert_eq!(VALUE.get(), Some(2));
        });

        assert_eq!(VALUE.get(), Some(1));
        handle.join().unwrap();

        assert_eq!(VALUE.get(), Some(1));
    }

    #[tokio::test]
    async fn test_future_once_lock() {
        static VALUE: FutureOnceLock<u64> = FutureOnceLock::new();

        let fut_1 = async {
            for _ in 0..42 {
                let j = VALUE.with(Clone::clone);
                VALUE.replace(j + 1);
                tokio::task::yield_now().await;
            }

            VALUE.get().unwrap()
        }
        .with_scope(&VALUE, 0);

        let fut_2 = async { VALUE.get().unwrap() }.with_scope(&VALUE, 15);

        assert_eq!(fut_1.await, 42);
        assert_eq!(fut_2.await, 15);
        assert_eq!(
            tokio::spawn(async { VALUE.get().unwrap() }.with_scope(&VALUE, 115))
                .await
                .unwrap(),
            115
        );
    }
}


examples/context/mod.rs
use std::{
    cell::RefCell,
    fmt::Display,
    future::Future,
    time::{Duration, Instant},
};

use future_local_storage::{FutureLocalStorage, FutureOnceCell};

/// Tracing span context.
static CONTEXT: FutureOnceCell<TracerContext> = FutureOnceCell::new();

#[derive(Debug)]
pub struct TracerContext {
    begin: Instant,
    traces: RefCell<Vec<TraceEntry>>,
}

impl TracerContext {
    /// Adds an "entered" event to the trace.
    pub fn on_enter(message: impl Display) {
        Self::with(|tracer| tracer.add_entry("entered", message));
    }

    /// Adds an "exit" event to the trace.
    pub fn on_exit(message: impl Display) {
        Self::with(|tracer| tracer.add_entry("exited", message));
    }

    /// Each future has its own tracing context in which it is executed, and after execution
    /// is complete, all events are returned along with the future output.
    pub async fn in_scope<R, F>(future: F) -> (Vec<TraceEntry>, R)
    where
        F: Future<Output = R>,
    {
        let (this, result) = future.with_scope(&CONTEXT, Self::new()).await;
        (this.traces.take(), result)
    }

    fn new() -> Self {
        let tracer = Self {
            begin: Instant::now(),
            traces: RefCell::new(Vec::new()),
        };
        tracer.add_entry("created", "a new async tracer started");
        tracer
    }

    fn with<R, F: FnOnce(&Self) -> R>(scope: F) -> R {
        CONTEXT.with(|tracer| scope(tracer))
    }

    fn add_entry(&self, event: impl Display, message: impl Display) {
        self.traces.borrow_mut().push(TraceEntry {
            duration_since: self.elapsed(),
            event: event.to_string(),
            message: message.to_string(),
        });
    }

    fn elapsed(&self) -> Duration {
        Instant::now().duration_since(self.begin)
    }
}

#[derive(Debug)]
pub struct TraceEntry {
    pub duration_since: Duration,
    pub event: String,
    pub message: String,
}


examples/async_tracer.rs
//! A toy async tracer example.

use std::{fmt::Display, time::Duration};

use context::{TraceEntry, TracerContext};

mod context;

impl Display for TraceEntry {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "{:>8.2}ms: {:>12} {}",
            self.duration_since.as_secs_f64() * 1_000_f64,
            self.event,
            self.message
        )
    }
}

async fn second_long_computation(a: u64) -> u64 {
    TracerContext::on_enter(format!("`second_long_computation` with params: a={a}"));

    // Imitate a CPU-intensive computation.
    for _ in 0..10 {
        std::thread::sleep(Duration::from_millis(10));
        tokio::task::yield_now().await;
    }

    TracerContext::on_exit("`second_long_computation`");
    a * 32
}

async fn first_computation(a: u64) -> u64 {
    TracerContext::on_enter(format!("`first_computation` with params: a={a}"));

    tokio::task::yield_now().await;

    TracerContext::on_exit("`first_computation`");
    a + 2
}

async fn some_method(name: impl Display, mut a: u64) -> u64 {
    tokio::task::yield_now().await;
    TracerContext::on_enter(format!("`some_method` with params: name='{name}', a={a}"));

    a = first_computation(a).await;
    a = second_long_computation(a).await;

    TracerContext::on_exit("`some_method`");
    a
}

#[tokio::main]
async fn main() {
    // Spawn a lot of async computations in the multithreading runtime.
    let handles: Vec<_> = [
        (1, "lorem"),
        (2, "ipsum"),
        (3, "dolor"),
        (4, "sit"),
        (5, "amet"),
        (6, "consectetur"),
        (7, "adipiscing"),
        (8, "elit"),
        (9, "mauris"),
        (10, "at consequat"),
        (11, "dui"),
        (12, "vel"),
        (13, "convallis"),
        (14, "purus"),
    ]
    .into_iter()
    .map(|(i, method)| tokio::spawn(TracerContext::in_scope(some_method(method, i))))
    .collect();

    // Wait for them for complete.
    let results = futures_util::future::join_all(handles).await;
    for result in results {
        let (trace, answer) = result.expect("unable to join future");

        println!("Computation finished with answer: {answer}");
        println!("Captured traces:");
        for entry in trace {
            println!("{entry}");
        }
        println!();
    }
}


src/lib.rs
//! # Overview
#![doc = include_utils::include_md!("README.md:description")]
//!
//! ## Examples
//!
//! ### Tracing spans
//!
//! ```rust
#![doc = include_str!("../examples/context/mod.rs")]
//!
//! // Usage example
//!
//! async fn some_method(mut a: u64) -> u64 {
//!     TracerContext::on_enter(format!("`some_method` with params: a={a}"));
//!    
//!     // Some async computation
//!     
//!     TracerContext::on_exit("`some_method`");
//!     a * 32
//! }
//!
//! #[tokio::main]
//! async fn main() {
//!     let (trace, result) = TracerContext::in_scope(some_method(45)).await;
//!
//!     println!("answer: {result}");
//!     println!("trace: {trace:#?}");
//! }
//! ```

use std::{fmt::Debug, future::Future};

use future::ScopedFutureWithValue;
use imp::FutureLocalKey;

pub mod future;
mod imp;

/// An init-once-per-future cell for thread-local values.
///
/// It uses thread local storage to ensure that the each polled future has its own local storage key.
/// Unlike the [`std::thread::LocalKey`] this cell will *not* lazily initialize the value on first access.
/// Instead, the value is first initialized when the future containing the future-local is first polled
/// by an executor.
///
/// After the execution finished the value moves from the future local cell to the future output.
pub struct FutureOnceCell<T>(imp::FutureLocalKey<T>);

impl<T> FutureOnceCell<T> {
    /// Creates an empty future once cell.
    #[must_use]
    pub const fn new() -> Self {
        Self(imp::FutureLocalKey::new())
    }
}

impl<T: Send + 'static> FutureOnceCell<T> {
    /// Acquires a reference to the value in this future local storage.
    ///
    /// Unlike the [`std::thread::LocalKey::with`] this method does not initialize the value
    /// when called.
    ///
    /// # Panics
    ///
    /// - This method will panic if the future local doesn't have a value set.
    ///
    /// - If you the returned future inside the a call to [`Self::with`] on the same cell, then the
    ///   call to `poll` will panic.
    #[inline]
    pub fn with<F, R>(&'static self, f: F) -> R
    where
        F: FnOnce(&T) -> R,
    {
        let value = self.0.local_key().borrow();
        f(value
            .as_ref()
            .expect("cannot access a future local value without setting it first"))
    }

    /// Returns a copy of the contained value.
    ///
    /// # Panics
    ///
    /// This method will panic if the future local doesn't have a value set.
    #[inline]
    pub fn get(&'static self) -> T
    where
        T: Copy,
    {
        self.0.local_key().borrow().unwrap()
    }

    /// Sets a value `T` as the future-local value for the future `F`.
    ///
    /// On completion of `scope`, the future-local value will be returned by the scoped future.
    ///
    /// ```rust
    /// use std::cell::Cell;
    ///
    /// use future_local_storage::FutureOnceCell;
    ///
    /// static VALUE: FutureOnceCell<Cell<u64>> = FutureOnceCell::new();
    ///
    /// #[tokio::main]
    /// async fn main() {
    ///     let (output, answer) = VALUE.scope(Cell::from(0), async {
    ///         VALUE.with(|x| {
    ///             let value = x.get();
    ///             x.set(value + 1);
    ///         });
    ///
    ///         42
    ///     }).await;
    /// }
    /// ```
    #[inline]
    pub fn scope<F>(&'static self, value: T, future: F) -> ScopedFutureWithValue<T, F>
    where
        F: Future,
    {
        future.with_scope(self, value)
    }
}

impl<T: Debug + Send + 'static> Debug for FutureOnceCell<T> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_tuple("FutureOnceCell").field(&self.0).finish()
    }
}

impl<T> AsRef<FutureLocalKey<T>> for FutureOnceCell<T> {
    fn as_ref(&self) -> &FutureLocalKey<T> {
        &self.0
    }
}

/// Attaches future local storage values to a [`Future`].
///
/// Extension trait allowing futures to have their own static variables.
pub trait FutureLocalStorage: Future + Sized + private::Sealed {
    /// Sets a given value as the future local value of this future.
    ///
    /// Each future instance will have its own state of the attached value.
    ///
    /// ```rust
    /// use std::cell::Cell;
    ///
    /// use future_local_storage::{FutureOnceCell, FutureLocalStorage};
    ///
    /// static VALUE: FutureOnceCell<Cell<u64>> = FutureOnceCell::new();
    ///
    /// #[tokio::main]
    /// async fn main() {
    ///     let (output, answer) = async {
    ///         VALUE.with(|x| {
    ///             let value = x.get();
    ///             x.set(value + 1);
    ///         });
    ///         
    ///         42
    ///     }.with_scope(&VALUE, Cell::from(0)).await;
    /// }
    /// ```
    fn with_scope<T, S>(self, scope: &'static S, value: T) -> ScopedFutureWithValue<T, Self>
    where
        T: Send,
        S: AsRef<FutureLocalKey<T>>;
}

mod private {
    use std::future::Future;

    pub trait Sealed {}

    impl<F: Future> Sealed for F {}
}

#[cfg(test)]
mod tests {
    use std::cell::{Cell, RefCell};

    use pretty_assertions::assert_eq;

    use super::*;
    use crate::FutureLocalStorage;

    #[test]
    fn test_once_cell_without_future() {
        static LOCK: FutureOnceCell<RefCell<String>> = FutureOnceCell::new();
        LOCK.0
            .local_key()
            .borrow_mut()
            .replace(RefCell::new("0".to_owned()));

        assert_eq!(LOCK.with(|x| x.borrow().clone()), "0".to_owned());
        LOCK.with(|x| x.replace("42".to_owned()));
        assert_eq!(LOCK.with(|x| x.borrow().clone()), "42".to_owned());
    }

    #[tokio::test]
    async fn test_future_once_cell_output() {
        static VALUE: FutureOnceCell<Cell<u64>> = FutureOnceCell::new();

        let (output, ()) = VALUE
            .scope(Cell::from(0), async {
                VALUE.with(|x| {
                    let value = x.get();
                    x.set(value + 1);
                });
            })
            .await;

        assert_eq!(output.into_inner(), 1);
    }

    #[tokio::test]
    async fn test_future_once_cell_discard_value() {
        static VALUE: FutureOnceCell<Cell<u64>> = FutureOnceCell::new();

        let fut_1 = async {
            for _ in 0..42 {
                VALUE.with(|x| {
                    let value = x.get();
                    x.set(value + 1);
                });
                tokio::task::yield_now().await;
            }

            VALUE.with(Cell::get)
        }
        .with_scope(&VALUE, Cell::new(0))
        .discard_value();

        let fut_2 = async { VALUE.with(Cell::get) }
            .with_scope(&VALUE, Cell::new(15))
            .discard_value();

        assert_eq!(fut_1.await, 42);
        assert_eq!(fut_2.await, 15);
        assert_eq!(
            tokio::spawn(
                async { VALUE.with(Cell::get) }
                    .with_scope(&VALUE, Cell::new(115))
                    .discard_value()
            )
            .await
            .unwrap(),
            115
        );
    }
}
